### PR TITLE
Handle granular requests configuration

### PR DESCRIPTION
**Pull request by snutij**
### Motivation

First iteration for: #834, I will follow up with client implementation after this one get validation. 

Allow user to configure `inlayHints` and `codeLens` requests, and disabling unwanted features. 

### Implementation

Handle in the server a hash of feature configuration:
- we set a default configuration so we could merge this without client implementation   
- merge client configurations with default ones
- early return in requests in case of disabled features

Question: should I disable features by default as proposed [here](https://github.com/Shopify/ruby-lsp/issues/834#issuecomment-1830853866)? To my very personal opinion, this can bring confusion with [supported features](https://shopify.github.io/ruby-lsp/RubyLsp/Requests.html) documentation. 
 
### Automated Tests

Tests added: 
- in `executor` to ensure default and provided configurations are well populated
- in both request to ensure there are no responses in case of disabled feature

### Manual Tests

By switching default configuration to `false` in `lib/ruby_lsp/store.rb`, then restarting, no feature should be display to user. 

<details><summary>Screenshots</summary>

#### inlayHints with `true` / `false` default configuration:
image
image

#### codelens with `true` / `false` default configuration: 
image
image
</details>

### CODE REVIEW

**Comment by vinistock** [comment]
This is amazing. Thank you for the contribution, Justin!

> Question: should I disable features by default as proposed https://github.com/Shopify/ruby-lsp/issues/834#issuecomment-1830853866? To my very personal opinion, this can bring confusion with [supported features](https://shopify.github.io/ruby-lsp/RubyLsp/Requests.html) documentation.

You raise a really good point. We discussed this with the team and I think we can do the following:

- Let's default all inlay hints to disabled
- Let's keep code lens as enabled by default
- Let's update both the documentation for code lens and inlay hints to mention these configurations
- Finally, let's add an `enableAll` entry to both configuration hashes. That way, someone interested in having the hints can simply turn everything on instead of needing to add every single one

In terms of code style, it might be worth creating an object in `utils.rb` to encapsulate the configuration logic. Something along these lines

```ruby
class RequestConfig
  def initialize(hash)
    @hash = hash
  end

  def enabled?(feature)
    @hash[:enableAll] || @hash[feature]
  end
end
```

**Review by vinistock — APPROVED** [review]
This is great! I think we're ready to merge, I just wanted to check if we need to return the configured features back to the client. If not, let's remove that line and merge

File: lib/ruby_lsp/executor.rb [hunk 1]
```diff
@@ -743,6 +750,7 @@ def initialize_request(options)
           version: VERSION,
         },
         formatter: @store.formatter,
+        features_configuration: @store.features_configuration,
```
[the remark this answers is not in the document]
Comment (snutij) [reply lib/ruby_lsp/executor.rb]: It was only for debugging purpose, to provide easier support on issue of kind "feature X is not working". Removed until we need it. 👍

**Comment by vinistock** [comment]
🚀

**Pull request merged**